ftp.go

```diff
@@ -84,11 +87,29 @@ func Dial(addr string, options ...DialOption) (*ServerConn, error) {
 		do.location = time.UTC
 	}
 
+	// Use the resolved IP address in case addr contains a domain name
+	// If we use the domain name, we might not resolve to the same IP.
+	addr, remoteAddr, err := resolveAddr(addr)
+	if err != nil {
+		return nil, err
+	}
+
 	tconn := do.conn
 	if tconn == nil {
 		var err error
 
-		if do.dialFunc != nil {
+		socksProxy := getSocksProxy(do)
+		if len(socksProxy) > 0 {
+			socksDialer, err := proxy.SOCKS5("tcp", socksProxy, nil, proxy.Direct)
```

Comment:
This block is redundant. Please create a separate function to avoid duplicate code.